### DISCUSSION TITLE
Reading RMQ source code: Please suggest one key feature to study deeply  请推荐一个RMQ功能进行研究

### DISCUSSION THREAD

**Discussion by vitasworld** [apache/rocketmq#8828] [Q&A] [answered]
这是我的问题摘要。请建议代码的一部分或一个功能，以便开始我的学习。我会阅读这部分代码，以提高我对RocketMQ至少一部分的理解。然后我会教我的同事我学到的东西。

--

Hello. I would like suggestions of which key feature(s) of RocketMQ  could be studied in the codebase that will help me develop a deeper understanding of the product.

You may be asking why I want to read the code rather than watching videos or reading docs.   This is what my employer wants.   I already have a basic understanding of RocketMQ.  My employer wants me to dive deeper, become a subject matter expert, and then share my learnings with my teammates.

I have agreed to read the code that implements a single key feature.  I have about a month to do this.  If anybody has suggestions about a part of the codebase that would be a good place to start, I would like to hear your opinion.

**Comment by leizhiyuan** [comment, marked as answer]
As a core feature of rocketmq5.0, maybe you can have an in-depth understanding of the pop consumption model and related rip

https://lists.apache.org/thread/0bm0qfj8xvqt1jndcmo6b76x6v2ptz1d

pop entry point for server-side consumption code
```
org.apache.rocketmq.proxy.grpc.v2.consumer.ReceiveMessageActivity receive
org.apache.rocketmq.proxy.grpc.v2.consumer.AckMessageActivity ack
```
Client, you can run this to trigger consumption so that you can debug on the server.

```
https://github.com/apache/rocketmq-clients/blob/master/java/client/src/main/java/org/apache/rocketmq/client/java/example/SimpleConsumerExample.java

```

**Reply by vitasworld** [reply]
Perfect thank you!
